Mjml.Net/MjmlRenderContext.cs

```diff
@@ -62,19 +55,25 @@ public ValidationErrors Validate()
             return validator?.Complete() ?? new ValidationErrors();
         }
 
-        public void ReadFragment(TextReader mjml)
+        public void ReadFragment(string mjml)
         {
-            using (var xml = XmlReader.Create(mjml, ReaderSettings))
+            if (mjmlOptions.Lax)
             {
-                ReadXml(xml, currentComponent);
+                mjml = XmlFixer.FixXML(mjml, mjmlOptions);
             }
-        }
 
-        public void ReadFragment(string mjml)
-        {
-            using (var xml = XmlReader.Create(new StringReader(mjml), ReaderSettings))
+            var fragmentReader = new XmlTextReader(mjml, XmlNodeType.Element, mjmlOptions.ParserContext)
             {
-                ReadXml(xml, currentComponent);
+                // Parse the doctype definition for the allowed entities.
+                DtdProcessing = DtdProcessing.Parse,
+
+                // Keep the entities.
+                EntityHandling = EntityHandling.ExpandCharEntities
```

Comment:
Since we're setting `DtdProcessing` to `Parse` we need to weary of security issues that can come with it. One of the security issues is a DoS (Denial of Service) attack e.g. The Billion Laughs Attack. Which is attack agains the XML parser and how it handles entity character expansion.

Here is an example:
```xml
<?xml version="1.0"?>
<!DOCTYPE lolz [
<!ENTITY lol "lol">
<!ENTITY lol2 "&lol;&lol;&lol;&lol;&lol;&lol;&lol;&lol;&lol;&lol;">
<!ENTITY lol3 "&lol2;&lol2;&lol2;&lol2;&lol2;&lol2;&lol2;&lol2;&lol2;&lol2;">
<!ENTITY lol4 "&lol3;&lol3;&lol3;&lol3;&lol3;&lol3;&lol3;&lol3;&lol3;&lol3;">
<!ENTITY lol5 "&lol4;&lol4;&lol4;&lol4;&lol4;&lol4;&lol4;&lol4;&lol4;&lol4;">
<!ENTITY lol6 "&lol5;&lol5;&lol5;&lol5;&lol5;&lol5;&lol5;&lol5;&lol5;&lol5;">
<!ENTITY lol7 "&lol6;&lol6;&lol6;&lol6;&lol6;&lol6;&lol6;&lol6;&lol6;&lol6;">
<!ENTITY lol8 "&lol7;&lol7;&lol7;&lol7;&lol7;&lol7;&lol7;&lol7;&lol7;&lol7;">
<!ENTITY lol9 "&lol8;&lol8;&lol8;&lol8;&lol8;&lol8;&lol8;&lol8;&lol8;&lol8;">
]>
<lolz>&lol9;</lolz>
```

We should set the `XmlReaderSettings.MaxCharactersFromEntities` to a default of `1024` to prevent this sort of attack by default. I think we should then surface this as an option via MjmlOptions.

I'll need to do further testing for XXE attacks.

Reply:
But we are setting the dtd ourselve. If the actually mjml would have another dtd then you will get an error. something like "duplicate dtd"

Reply:
Great, as long as we're in control of the DTD then that's perfect. However, we should add in the mitigation steps in case we ever introduced a bug or functionality that could allow malicious control of the constructed DTD.

Reply:
How would you do that? You need the dtd parsing to make the entities known. The whole task does not work otherwise.

Reply:
I was wrong. I thought that the xml context acts as a dtd, but it does not. So I have set a dummy dtd to prevent the custom dtd. I created a test for that (see SecurityTests.cs)

Reply:
I can think of one scenario. 

If a user of the library allowed custom control of the XmlEntities dictionary. It could be possible to inject a XML bomb. 

```csharp
// Where we build the DTD based upon XmlEntites
sb.AppendLine(CultureInfo.InvariantCulture, $"!ENTITY {key[1..]} \"{value}\">");
```
For example, this is a valid DTD and using this input dictionary of:
- SafeEntityDefinition = "233"
```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE example [
<!ENTITY SafeEntityDefinition "233">
]>
<example>&SafeEntityDefinition;</example>
```

Now, if an malicious actor can control the known entities they could do something like this in the dictionary:
- SafeEntityDefinition = "233"
- UnsafeEntityDefinition = `\"><!ENTITY BadInjectedDefinition \"XML BOMB ENTITIES\">`

This would be injected into the DTD template like so:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE example [
<!ENTITY SafeEntityDefinition "233">
<!ENTITY SafeEntityDefinition ""><!ENTITY BadInjectedDefinition "XML BOMB ENTITIES">
]>
<example>&SafeEntityDefinition;</example>
```

Reply:
I am not sure if I understand that. The "XML BOMB ENTITIES" are never used, because the actually entities is kept in the output.

Reply:
I think ultimately, this is a decision on who should be responsible for prevention of these cases. The developer or the library! 😄

Reply:
I'll download the PR and see

Reply:
The The Billion Laughs Attack cannot happen (we could write a test for that, ofc).

Because replacement actually never happens. Thats the whole point of this PR, we keep &nbsp; and do not replace it with #1232; (or whatever it is).

Reply:
> * &BadInjectedDefinition;"><!ENTITY BadInjectedDefinition "XML BOMB ENTITIES

You're right. I misunderstood on my end, sorry. In that case, we're good to go.

Reply:
Now, I am confused. It turns out that with the setting EntityHandling.ExpandCharEntities you do not need a dtd anymore.

Reply:
> Now, I am confused. It turns out that with the setting EntityHandling.ExpandCharEntities you do not need a dtd anymore.

Strange, I was under the impression that a DTD is required for defining custom XML entities.

Reply:
I've just done a quick test and I don't think we need to set `EntityHandling.ExpandCharEntities`?

Reply:
Why not? This keeps the &nbsp; in the output.

Reply:
> Why not? This keeps the   in the output.

I've commented out Entity Handling in both `MjmlRendererContext` and `MjmlRenderer`.

This is the test:

```html
<mjml>
    <mj-head></mj-head>
    <mj-body>
        <mj-text><span>hello &cent; world            </span></mj-text>
    </mj-body>
</mjml>
```

```csharp
Assert.Contains("<span>hello &cent; world            </span>", result.Html!, StringComparison.OrdinalIgnoreCase);
```

and all the tests pass for me including this one.

Reply:
I don't understand that. For my understanding HTML entities are just like variables. You say &abc; should be replaced with `abc-value`. But to make it work you have to make the html entities known, otherwise you would receive the exception that is currently described in the Readme.

So you have 2 options:

1. Make the entity known, but then you cannot keep the &abc; in the output because this would be a self referencing entitiy.
2. Tell the xml reader to basically keep the entities ,so that you can do the replacement itself.

In the other PR I have chosen option 2.

Reply:
Agreed, that was my understanding too. As when we've downloaded the sample MJML templates from mjml.io we received the unknown entity errors in the XmlReader.

It's my understanding that EntityHandling.ExpandEntityChars is only needed when a entity references another entity and needs to expand the result.

Which would suggest to me that the only requirement we need is DtdProcessing.Parse.

However, Im confused why this works without it!

Reply:
But why the hell have we then introduced the option?

Reply:
Edit: I have found it:



var rootReader2 = XmlReader.Create(new StringReader(xml));
// rootReader.EntityHandling = ExpandEntities;

var rootReader = new XmlTextReader(xmlRoot, XmlNodeType.Document, (XmlParserContext)null);
// rootReader.EntityHandling = ExpandCharEntities;

The default is just different. But I would recommend to make it explicit to be on the safe side.

Reply:
> Edit: I have found it:
> 
> var rootReader2 = XmlReader.Create(new StringReader(xml)); // rootReader.EntityHandling = ExpandEntities;
> 
> var rootReader = new XmlTextReader(xmlRoot, XmlNodeType.Document, (XmlParserContext)null); // rootReader.EntityHandling = ExpandCharEntities;
> 
> The default is just different. But I would recommend to make it explicit to be on the safe side.

Good catch, I agree.